### ISSUE TITLE
MsalClientException: The same command was already received and is being processed.

### ISSUE THREAD

**Issue by z3ntu** [AzureAD/microsoft-authentication-library-for-android#1101]
**Describe the bug**

In this app I have the authentication set up so that the current access token is added to the request headers, then when the request fails because of a 401 Unauthorized response we call `ISingleAccountPublicClientApplication.acquireTokenSilentAsync` to get a new token and then retry the request with this token. This is solved by using an OkHttp Interceptor that does that.
Unfortunately more often than not, when the current access token expires (which seems to be a couple of hours) the refresh fails with the exception `com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.`. This causes the user to be logged out of the app (this is how we handle unrecoverable 401 errors) so this is quite a serious issue in this app.

**Smartphone (please complete the following information):**
 - Device: OnePlus 6T (and more)
 - Android Version: Android 10
 - Browser Chrome
 - MSAL Version 1.5.1

**Stacktrace**
```
    com.microsoft.identity.client.exception.MsalClientException: The same command was already received and is being processed.
        at com.microsoft.identity.client.internal.controllers.MsalExceptionAdapter.msalExceptionFromBaseException(MsalExceptionAdapter.java:51)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:339)
        at com.microsoft.identity.client.SingleAccountPublicClientApplication$3.onError(SingleAccountPublicClientApplication.java:325)
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:93)
        at android.os.Handler.handleCallback(Handler.java:883)
        at android.os.Handler.dispatchMessage(Handler.java:100)
        at android.os.Looper.loop(Looper.java:214)
        at android.app.ActivityThread.main(ActivityThread.java:7682)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
     Caused by: com.microsoft.identity.common.exception.ClientException: The same command was already received and is being processed.
        at com.microsoft.identity.common.internal.controllers.CommandDispatcher$1.run(CommandDispatcher.java:93) 
        at android.os.Handler.handleCallback(Handler.java:883) 
        at android.os.Handler.dispatchMessage(Handler.java:100) 
        at android.os.Looper.loop(Looper.java:214) 
        at android.app.ActivityThread.main(ActivityThread.java:7682) 
        at java.lang.reflect.Method.invoke(Native Method) 
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:516) 
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:950)
```

**To Reproduce**
Probably calling `acquireTokenSilentAsync` a second time (on a different thread) before the first one completes.

**Expected behavior**
The `acquireTokenSilentAsync` function should return a new token without any error.

**Actual Behavior**
The `acquireTokenSilentAsync` function threw an exception in `onError`

Probably related to #847 but that one's closed.

**Comment by shoatman** [comment]
@z3ntu - We've introduced throttling to prevent duplicate requests from going to the server concurrently.  It sounds like that is what may be happening in your case.  Can you provide more information on your scenario?  Do you know why a second duplicate request is being started at the same time as the first?

**Comment by shoatman** [comment]
I'm marking this as a bug as I believe throttling is a behavioral breaking change and that we need to look at this more closedly.  Would still appreciate some additional context on your scenario though.

**Comment by z3ntu** [comment]
As I mentioned this is being called as part of an OkHttp interceptor which acts when a network call (where I normally place the cached Auth token into the header) is returning a HTTP 401 response. Then it calls the MSAL function to get a new auth token. During app start (where the old cached auth token likely expired - seems to be ~2 hours) multiple requests are sent at the same time which both get a response back at roughly the same time causing the MSAL function to be called twice on different threads (I think, not 100% sure about the internals) at more or less the same time.

**Comment by shoatman** [comment]
@shahzaibj - Here's the scenario for reference.

**Comment by narkusv** [comment]
We've been having same issue, but resolved it by wrapping call to get access token in a ReentrantLock 
`    lock.withLock {
            mSingleAccountApp.acquireTokenSilent(getScopes(), AUTHORITY_SIGN_IN_PROD).accessToken
        }`

**Comment by iambmelt** [comment]
I've been investigating how to cut-down on these errors served up by MSAL; I've submitted [this change](https://github.com/AzureAD/microsoft-authentication-library-common-for-android/pull/1012) which will roll-out in our next release. This should offer relief from these errors and give a boost to many-threaded scenarios by allowing result-sharing for duplicate_requests

**Comment by z3ntu** [comment]
As long as I'm getting a token back and not an error, I'm happy. :smile: 
Also FTR I also currently have a `synchronized(...) { }` block around the MSAL call to avoid this issue and it seems to work around that reliably.

**Comment by iambmelt** [comment]
Fixed in [2.0.1](https://github.com/AzureAD/microsoft-authentication-library-for-android/releases/tag/v2.0.1). Please reopen or file a new issue if you have any problems. Thanks!
